auth-service/src/main/kotlin/uk/gov/communities/delta/auth/services/GroupService.kt

```diff
@@ -45,19 +55,50 @@ class GroupService(
         groupCN: String,
         call: ApplicationCall,
         triggeringAdminSession: OAuthSession? = null,
+    ) {
+        addUserToGroup(adUser.cn, adUser.dn, groupCN, call, triggeringAdminSession)
+    }
+
+    suspend fun addUserToGroup(
+        userCN: String,
+        userDN: String,
+        groupCN: String,
+        call: ApplicationCall,
+        triggeringAdminSession: OAuthSession? = null,
     ) {
         val groupDN = ldapConfig.groupDnFormat.format(groupCN)
 
         if (!groupExists(groupDN)) {
             createGroup(groupCN)
         }
         ldapServiceUserBind.useServiceUserBind {
-            val member = BasicAttribute("member", adUser.dn)
+            val member = BasicAttribute("member", userDN)
             val modificationItems = arrayOf(ModificationItem(DirContext.ADD_ATTRIBUTE, member))
             it.modifyAttributes(groupDN, modificationItems)
-            logger.atInfo().addKeyValue("UserDN", adUser.dn).log("User added to group with dn {}", groupDN)
+            logger.atInfo().addKeyValue("UserDN", userDN).log("User added to group with dn {}", groupDN)
+        }
+        auditAddingUserToGroup(userCN, groupCN, triggeringAdminSession, call)
+    }
+
+    suspend fun removeUserFromGroup(
+        userCN: String,
+        userDN: String,
+        groupCN: String,
+        call: ApplicationCall,
+        triggeringAdminSession: OAuthSession? = null,
+    ) {
+        val groupDN = ldapConfig.groupDnFormat.format(groupCN)
+
+        if (!groupExists(groupDN)) {
+            throw Exception("Group did not exist in AD when trying to remove a user from it")
+        }
```

Comment:
Do we not get a nice error message without this?

Reply:
Its a NameNotFoundException which aren't the easiest to decipher but since they do actually have the information and this adds slowness I have removed these lines :)